hack/tools/janitor/janitor.go

```diff
@@ -0,0 +1,245 @@
+		dryRun:          dryRun,
+		ipamClient:      ipamClient,
+		ipamNamespace:   ipamNamespace,
+		maxCreationDate: time.Now().Add(-delta),
+		vSphereClients:  vSphereClients,
+	}
+}
+
+type janitor struct {
+	dryRun          bool
+	ipamClient      client.Client
+	ipamNamespace   string
+	maxCreationDate time.Time
+	vSphereClients  *vSphereClients
+}
+
+type virtualMachine struct {
+	managedObject mo.VirtualMachine
+	object        *object.VirtualMachine
+}
+
+// vSphereVMs deletes all vSphereVMs in a given folder in vSphere if their creation
+// timestamp is before the janitor's configured maxCreationDate.
+func (s *janitor) vSphereVMs(ctx context.Context, folder string) error {
+	log := ctrl.LoggerFrom(ctx).WithName("vSphereVMs").WithValues("folder", folder)
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	if folder == "" {
+		return fmt.Errorf("cannot use empty string as folder")
+	}
+
+	log.Info("Deleting vSphere VMs in folder")
+
+	// List all virtual machines inside the folder.
+	finder := find.NewFinder(s.vSphereClients.Vim, false)
+	managedObjects, err := finder.ManagedObjectListChildren(ctx, folder+"/...", "VirtualMachine")
+	if err != nil {
+		return err
+	}
+
+	if len(managedObjects) == 0 {
+		return nil
+	}
+
+	// Retrieve information for all found virtual machines.
+	managedObjectReferences := []types.ManagedObjectReference{}
+	for _, obj := range managedObjects {
+		managedObjectReferences = append(managedObjectReferences, obj.Object.Reference())
+	}
+	var managedObjectVMs []mo.VirtualMachine
+	if err := s.vSphereClients.Govmomi.Retrieve(ctx, managedObjectReferences, []string{"config", "summary.runtime.powerState"}, &managedObjectVMs); err != nil {
+		return err
+	}
+
+	vmsToDeleteAndPoweroff := []*virtualMachine{}
+	vmsToDelete := []*virtualMachine{}
+
+	// Filter out vms we don't have to cleanup depending on s.maxCreationDate.
+	for _, managedObjectVM := range managedObjectVMs {
+		if managedObjectVM.Config.CreateDate.After(s.maxCreationDate) {
+			// Ignore vms created after maxCreationDate
+			continue
+		}
+
+		vm := &virtualMachine{
+			managedObject: managedObjectVM,
+			object:        object.NewVirtualMachine(s.vSphereClients.Vim, managedObjectVM.Reference()),
+		}
+
+		if vm.managedObject.Summary.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOn {
+			vmsToDeleteAndPoweroff = append(vmsToDeleteAndPoweroff, vm)
+			continue
+		}
+		vmsToDelete = append(vmsToDelete, vm)
+	}
+
+	// PowerOff vms which are still running. Triggering PowerOff for a VM results in a task in vSphere.
+	poweroffTasks := []*object.Task{}
+	for _, vm := range vmsToDeleteAndPoweroff {
+		log.Info("Powering off vm in vSphere", "vm", vm.managedObject.Config.Name)
+		if s.dryRun {
+			// Skipping actual PowerOff on dryRun.
+			continue
+		}
+		task, err := vm.object.PowerOff(ctx)
+		if err != nil {
+			return err
+		}
+		poweroffTasks = append(poweroffTasks, task)
+	}
+	// Wait for all PowerOff tasks to be finished. We intentionally ignore errors here
```

Comment:
So this is overall a best effort powerOff so they give back their IPs? (let's add a godoc with some context around why we are trying poweroff)

Reply:
nope, thats a hard poweroff which is required before deleting a VM.

I think we're good with the dhcp lease duration :-)

Reply:
I meant best effort because we only log errors in l.138 :)

Reply:
So I think the open question is. Should we fail here or is it fine if we just log if an error occurs